### PR TITLE
KAFKA-482: Add change.data.capture.handler.remove.null.values to strip nulls in CDC write path

### PR DESCRIPTION
**Pull request by Calvinnix**

### CODE REVIEW

File: src/main/java/com/mongodb/kafka/connect/sink/cdc/CdcNullFieldValueRemover.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
```
Comment (andrewzlchen) [inline src/main/java/com/mongodb/kafka/connect/sink/cdc/CdcNullFieldValueRemover.java:2]: similar for the other new files

```suggestion
 * Copyright 2026-present MongoDB, Inc.
```

File: src/test/java/com/mongodb/kafka/connect/sink/MongoProcessedSinkRecordDataTest.java [hunk 2]
```diff
@@ -132,6 +133,108 @@ void testCDCHandling() {
     assertWriteModel(processedData, CDC_EXPECTED_WRITE_MODEL);
   }
 
+  @Test
+  @DisplayName("CDC null removal: default off leaves nulls in replacement")
+  void testCDCNullRemovalDefaultOff() {
+    String insertWithNulls = "{_id: 2, first_name: 'Bob', last_name: null, middle_name: null}";
+    String valueWithNulls =
+        format(
+            "{_id: 2, op: 'c', before: null, after: \"%s}\", source: 'ignored'}", insertWithNulls);
```
Comment (andrewzlchen) [inline src/test/java/com/mongodb/kafka/connect/sink/MongoProcessedSinkRecordDataTest.java]: hmm the tests seemed to have passed, but do we need the `}` in the `after` field?

```suggestion
            "{_id: 2, op: 'c', before: null, after: \"%s\", source: 'ignored'}", insertWithNulls);
```

File: src/test/java/com/mongodb/kafka/connect/sink/MongoProcessedSinkRecordDataTest.java [hunk 3]
```diff
@@ -132,6 +133,108 @@ void testCDCHandling() {
...
+  }
+
+  @Test
+  @DisplayName("CDC null removal: flag on strips nulls from CDC ReplaceOneModel")
+  void testCDCNullRemovalFlagOn() {
+    String insertWithNulls = "{_id: 3, first_name: 'Carol', last_name: null, middle_name: null}";
+    String expectedCleaned = "{_id: 3, first_name: 'Carol'}";
+    String valueWithNulls =
+        format(
+            "{_id: 3, op: 'c', before: null, after: \"%s}\", source: 'ignored'}", insertWithNulls);
```
Comment (andrewzlchen) [inline src/test/java/com/mongodb/kafka/connect/sink/MongoProcessedSinkRecordDataTest.java]:
```suggestion
            "{_id: 3, op: 'c', before: null, after: \"%s\", source: 'ignored'}", insertWithNulls);
```

File: src/main/java/com/mongodb/kafka/connect/sink/cdc/CdcNullFieldValueRemover.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+      BsonDocumentNullValueRemover.removeNullValues(replacement);
+      return replacement.isEmpty() ? null : model;
+    }
+    if (model instanceof UpdateOneModel) {
+      Bson update = ((UpdateOneModel<BsonDocument>) model).getUpdate();
+      if (!(update instanceof BsonDocument)) {
+        return model;
+      }
+      BsonDocument updateDoc = (BsonDocument) update;
+      BsonValue setVal = updateDoc.get(SET);
```
Comment (andrewzlchen) [inline src/main/java/com/mongodb/kafka/connect/sink/cdc/CdcNullFieldValueRemover.java:60]: does this need to handle more updateOne operations? ie: $inc, $addToSet, etc. ([update operators](https://www.mongodb.com/docs/manual/reference/mql/update/#std-label-update-operators))

Instead of grabbing the value of `$set`, I wonder if we can just assert that the key is one of the ones listed in the docs, and then walk the document like we do in `removeNullValues`

File: src/main/java/com/mongodb/kafka/connect/sink/cdc/CdcNullFieldValueRemover.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+      if (setVal != null && setVal.isDocument()) {
+        BsonDocument setDoc = setVal.asDocument();
+        BsonDocumentNullValueRemover.removeNullValues(setDoc);
+        if (setDoc.isEmpty()) {
+          updateDoc.remove(SET);
+        }
+      }
+      return updateDoc.isEmpty() ? null : model;
+    }
+    return model;
```
Comment (andrewzlchen) [inline src/main/java/com/mongodb/kafka/connect/sink/cdc/CdcNullFieldValueRemover.java:70]: q: just as a sanity check, do we need to handle DeleteOne?

File: src/test/java/com/mongodb/kafka/connect/sink/MongoProcessedSinkRecordDataTest.java [hunk 6]
```diff
@@ -132,6 +133,108 @@ void testCDCHandling() {
...
+  void testCDCNullRemovalWithoutCDCHandler() {
+    MongoProcessedSinkRecordData processedData =
+        new MongoProcessedSinkRecordData(
+            SINK_RECORD,
+            createSinkConfig(CHANGE_DATA_CAPTURE_HANDLER_REMOVE_NULL_VALUES_CONFIG, "true"));
+
+    assertWriteModel(processedData);
+  }
+
+  @Test
```
Comment (andrewzlchen) [inline src/test/java/com/mongodb/kafka/connect/sink/MongoProcessedSinkRecordDataTest.java:210]: can we add a test where a write model that looks like this `{foo: {$set: null}}` collapses to become a no-op?

**Review by andrewzlchen — COMMENTED** [review]
Looks great! I had a few questions, but once we address them, this will basically LGTM!

File: src/main/java/com/mongodb/kafka/connect/sink/cdc/CdcNullFieldValueRemover.java (hunk 1 above)
Comment (Calvinnix) [reply src/main/java/com/mongodb/kafka/connect/sink/cdc/CdcNullFieldValueRemover.java:2]: fwiw all copyright headers in this project have the same 2008-present wording 🤷‍♂️

File: src/main/java/com/mongodb/kafka/connect/sink/cdc/CdcNullFieldValueRemover.java (hunk 5 above)
Comment (Calvinnix) [reply src/main/java/com/mongodb/kafka/connect/sink/cdc/CdcNullFieldValueRemover.java:70]: No we don't, [DeleteOne](https://github.com/mongodb/mongo-java-driver/blob/r4.7.2/driver-core/src/main/com/mongodb/client/model/DeleteOneModel.java) just has a filter, no body so we wouldn't have anything to remove nulls from.

File: src/main/java/com/mongodb/kafka/connect/sink/cdc/CdcNullFieldValueRemover.java (hunk 4 above)
Comment (Calvinnix) [reply src/main/java/com/mongodb/kafka/connect/sink/cdc/CdcNullFieldValueRemover.java:60]: Hmm I'm pretty sure only $set or $unset are used to update documents for the connector.

File: src/test/java/com/mongodb/kafka/connect/sink/MongoProcessedSinkRecordDataTest.java (hunk 2 above)
Comment (andrewzlchen) [reply src/test/java/com/mongodb/kafka/connect/sink/MongoProcessedSinkRecordDataTest.java]: can you check to see if these tests are running as part of evergreen? im not sure why the tests pass in this case

File: src/main/java/com/mongodb/kafka/connect/sink/cdc/CdcNullFieldValueRemover.java (hunk 4 above)
Comment (andrewzlchen) [reply src/main/java/com/mongodb/kafka/connect/sink/cdc/CdcNullFieldValueRemover.java:60]: Ah yeah that checks out [here](https://github.com/mongodb/mongo-kafka/blob/c623ec0c1382eb34ea43c25288434c47acc01fd9/src/main/java/com/mongodb/kafka/connect/sink/cdc/mongodb/operations/OperationHelper.java#L150-L166)

thanks for confirming

**Review by andrewzlchen — APPROVED** [review]
lgtm!
